pkg/ddl/partition.go

```diff
@@ -3869,54 +3837,65 @@ func (w *reorgPartitionWorker) BackfillData(handleRange reorgBackfillTask) (task
 		taskCtx.nextKey = nextKey
 		taskCtx.done = taskDone
 
-		isClustered := w.reorgedTbl.Meta().IsCommonHandle || w.reorgedTbl.Meta().PKIsHandle
-		if !isClustered {
-			// non-clustered table, we need to replace the _tidb_rowid handles since
-			// there may be duplicates across different partitions, due to EXCHANGE PARTITION.
-			// Meaning we need to check here if a record was double written to the new partition,
-			// i.e. concurrently written by StateWriteOnly or StateWriteReorganization.
-			// and we should skip it.
-			var found map[string][]byte
+		failpoint.InjectCall("PartitionBackfillData", len(w.rowRecords) > 0)
+		// For non-clustered tables, we need to replace the _tidb_rowid handles since
+		// there may be duplicates across different partitions, due to EXCHANGE PARTITION.
+		// Meaning we need to check here if a record was double written to the new partition,
+		// i.e. concurrently written by StateWriteOnly or StateWriteReorganization.
+		// and if so, skip it.
+		var found map[string][]byte
+		if len(w.oldKeys) > 0 {
+			// we must check if old IDs already been written,
+			// i.e. double written by StateWriteOnly or StateWriteReorganization.
+			// TODO: while waiting for BatchGet to check for duplicate, do another round of reads in parallel?
+			found, err = txn.BatchGet(ctx, w.oldKeys)
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+
+		for i, prr := range w.rowRecords {
+			taskCtx.scanCount++
+			key := prr.key
+
 			if len(w.oldKeys) > 0 {
-				// we must check if old IDs already been written,
-				// i.e. double written by StateWriteOnly or StateWriteReorganization.
-				// The good thing is that we can then also skip the index generation for that row and we don't need to
-				// check if duplicate index entries was already copied either!
-				// TODO: while waiting for BatchGet to check for duplicate, do another round of reads in parallel?
-				found, err = txn.BatchGet(ctx, w.oldKeys)
+				if _, ok := found[string(w.oldKeys[i])]; ok {
+					// Already filled, i.e. double written earlier by concurrent DML
+					continue
+				}
+
+				// Check if we can write the old key, since there can still be concurrent update
+				// happening on the rows from fetchRowColVals(), we can involve them in this
+				// transaction and fail when committing, so we will retry with a new snapshot.
+				// TODO: Test a concurrent pessimistic transaction that reads/writes concurrently here,
+				// and commits after after this transaction.
+				err = txn.Set(w.oldKeys[i], prr.vals)
+				if err != nil {
+					return errors.Trace(err)
+				}
+				// Ensure the key does not exist!
+				// It actually works without this assertion, but it may help troubleshooting.
+				err = txn.SetAssertion(w.oldKeys[i], kv.SetAssertNotExist)
+				if err != nil {
+					return errors.Trace(err)
+				}
+				// Also don't actually write it :)
+				err = txn.Delete(w.oldKeys[i])
```

Comment:
I'm not sure how to only force the transaction to fail if a key already exists, regardless of settings like `tidb_txn_assertion_level`, so by involving the key in the transaction, it will fail if another concurrent transaction has modified it (like insert it due to UPDATE). Just having the `SetAssertion()` is not enough, we could have the `txn.Delete()` or `txn.Set()` only, not sure if Set+Assert+Delete is better though, what is your thought?

Reply:
I don't quite understand this code. But why does `SetAssertNotExist` succeed after `txn.Set` it first?

Reply:
I would expect that only lock calls are directly checked/forwarded to the KV store, while SetAssertion/Set/Delete is only applied during Commit.

Reply:
If the goal is simply to prevent the row_key from being modified, using txn.LockKeys to apply a pessimistic lock on the corresponding row is the appropriate approach.

The purpose of Assertion is fundamentally different:
- It is designed to validate invariants or constraints to ensure correctness is not violated.
- It is not intended for concurrency control or preventing concurrent modifications.

Using LockKeys in a pessimistic transaction explicitly handles concurrency by preventing conflicting writes, while assertions act as safeguards to check assumptions about the state after operations are performed.

Reply:
For the use of internal transactions during the execution of DDL tasks and their concurrency control with DML, it is recommended to consult DDL-related colleagues to confirm whether the logic complies with DDL constraint requirements.

There should be similar code references for regular DDL backfill.
/cc @wjhuang2016 @tangenta

Reply:
The difference with regular DDL it uses the version/lock on a key to control another unrelated key. Looks too strange here.

Reply:
After changing to `txn.LockKeys()` it is now same as what `(*addIndexTxnWorker) BackfillData()` does [here](https://github.com/pingcap/tidb/blob/0be1983389d04348f72de36c4d922c333389c1e4/pkg/ddl/index.go#L2327).